client: fix edge cases in fscrypt truncate

chrisphoffman opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

chrisphoffman committed:
client: zero pad fscrypt last block

Before encrypting fscrypt last block, zero pad up to
FSCRYPT_BLOCK_SIZE.

Fixes: https://tracker.ceph.com/issues/73497
Signed-off-by: Christopher Hoffman <choffman@redhat.com>

chrisphoffman committed:
client: explicitly deny fscrypt fallocate

Do not allow fallocate if inode is encrypted. This operation if
implemented would require a first and/or last block RMW if offset
and/or offset+length is not aligned to fscrypt block size
(ie punch_hole). This sys call should not be modifying actual
data so we skip and do not support it.

This matches functionality from the kernel client:
https://github.com/torvalds/linux/blob/3f9f0252130e7dd60d41be0802bf58f6471c691d/fs/ceph/file.c#L2676-L2677

Fixes: https://tracker.ceph.com/issues/73497
Signed-off-by: Christopher Hoffman <choffman@redhat.com>

gregsfortytwo commented on qa/tasks/cephfs/test_fscrypt.py:
```diff
@@ -106,6 +106,15 @@ class TestFSCryptRMW(FSCryptTestCase):
+                'auth', 'caps', "client.1",
+                'mds', 'allow *',
+                'mon', 'allow *',
+                'osd', 'allow *')
```
Do we not have a client profile that works for these tests?

gregsfortytwo commented on src/client/Client.cc:
```diff
@@ -8621,14 +8627,13 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
 
       C_SaferCond *io_finish_cond = nullptr;
-      io_finish_cond = new C_SaferCond("Client::_read_async flock");
+      io_finish_cond = new C_SaferCond("Client::_do_setattr flock");
```
😮

chrisphoffman commented on src/client/Client.cc:
```diff
@@ -17487,6 +17487,9 @@ int Client::_fallocate(Fh *fh, int mode, int64_t offset, int64_t length)
 
+  if (in->is_fscrypt_enabled())
+    return -EOPNOTSUPP;
+
```
Yes disabling it at the test itself is possible

```
./fsx --help | grep fallocate
-F: Do not use fallocate (preallocation) calls
```

I've tried to keep the implementation best as I can to kernel spec. Do we want to deviate here?

chrisphoffman committed:
mds: wait until last block persisted to return

The implementation of truncating smaller of an fscrypt
file will require the use of fscrypt last block. This
process is a multi-step process. There is a race condition
that libcephfs can hit.

The problem:
1. Truncate an fscrypt file smaller
2. Immediately read file
3. Truncate larger
4. Read file

The expected contents of the file when step 4 is reached
is file contents truncated to smaller size and then zero
padded. The actual contents show the before contents of
truncating smaller (step 1).

This is due to a race condition due to the mds reporting back
to the client early before the last block is written persistently.
In the above example stale data will be read.

Why does this only exist in userspace?
The stale data read behavior is only present in libcephfs. This is
to do with how cache invalidation works. The kclient stores data in
the page cache in encrypted or decrypted form. Alternatively,
libcephfs may rely on ObjectCacher to cache data and stores it in
encrypted form.

During the above scenario, the kclient will truncate the decrypted
data and no read from osd is needed. Alternatively, libcephfs needs
to invalidate the complete block and will trigger a read from the osds.

Fixes: https://tracker.ceph.com/issues/74883
Signed-off-by: Christopher Hoffman <choffman@redhat.com>

chrisphoffman committed:
client: do not mix read cache modes during fscrypt truncates

During an fscrypt truncate, use correct read (oc read or read_trunc)
depending on client settings for preparing the last block.

Signed-off-by: Christopher Hoffman <choffman@redhat.com>

chrisphoffman commented:
After further testing, `fsx.sh` is still failing. The failures seem to happen less now, but still present. I'm guessing `no_early_reply = true` shrinks the window of the race condition, but doesn't eliminate it.

The cause of this conditions involves when the last block is written to the mds log but not flushed to the osds before the client attempts a sync read (cache miss). See this tracker for more details[1]. While the proposed fix narrows this window it doesn't fix it and thus this PR/issue needs further work.

Possible paths forward:
1. Do not release xlock on filelock until osd is completed
2. Expand on the functionality of `no_early_reply = true` to not return until fscrypt truncate smaller payload is flushed completely to the data pool on the osds.

I'm marking this as draft and will change back once a suitable complete fix is worked out.

1-https://tracker.ceph.com/issues/74883#note-7

chrisphoffman committed:
client: make sure to fsync and release caps during last block

Signed-off-by: Christopher Hoffman <choffman@redhat.com>

chrisphoffman committed:
client: Invalidate stale fscrypt data on smaller truncate

Fixes: https://tracker.ceph.com/issues/73497
Signed-off-by: Christopher Hoffman <choffman@redhat.com>

chrisphoffman commented:
>Expand on the functionality of no_early_reply = true to not return until fscrypt truncate smaller payload is flushed completely to the data pool on the osds.

I took a version of this. While I didn't change `no_early_reply` to mean something else, the server does not respond to the client until after last block is written.

@gregsfortytwo this is ready for review, can you take a look?
